### PR TITLE
Script for pinning ipfs data 

### PR DESCRIPTION
**Pull request by rohitverma007**
### Description:

Implements #1086 

### Checklist:

- [x] Test your work and double-check to confirm that you didn't break anything
- [ ] Wrap any displayed ETH addresses with [`formattedAddress`](https://github.com/OriginProtocol/origin/blob/master/origin-dapp/src/utils/user.js#L15-L17)
- [ ] Wrap any new text/strings for translation
- [ ] Map any new environment variables with a default value in the Webpack config
- [ ] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/origin-docs)

### CODE REVIEW

File: deployment/utils/gcf-pin-ipfs-hash/index.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    : null;
+
+  let pinnedHashes = [];
+  let unPinnedHashes = [];
+
+  if (!Object.is(data, null)) {
+    const hashesToPin = parseIncomingData(data);
+    for(let i = 0; i < hashesToPin.length; i++){
+      hashToPin = hashesToPin[i];
+      pinned = await ipfsClusterApiService.pin(hashToPin);
```
Comment (franckc) [inline deployment/utils/gcf-pin-ipfs-hash/index.js]: Should we have a reasonable number of retries before giving up ? Perhaps with some exponential delay.

File: deployment/utils/gcf-pin-ipfs-hash/index.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  } else {
+    console.log("Error retreiving listing data");
+  }
+  console.log("Pinned following hashes: ", pinnedHashes.join(", "), "\n");
+  console.log("Could not pin following hashes: ", unPinnedHashes.join(", "), "\n");
+};
+
+const parseIncomingData = data => {
+  let hashesToPin = [];
+  if(data.type == "listing"){
```
Comment (franckc) [inline deployment/utils/gcf-pin-ipfs-hash/index.js]: super nit: in our code base we try to always use the === operator.
Same comment applies to the other places where == is used in this PR.

**Comment by tomlinton** [comment]
Looks great @rohitverma007. Nothing to add aside from what @franckc has said.

I'm happy with where it lives along side the other Google Cloud function we have. The functionality is a bit niche and I don't think we want a full blown package. Also GCF needs this specific non-package structure if I remember correctly.

**Comment by rohitverma007** [comment]
@tomlinton @franckc 

I forgot to mention one key detail I just remembered as I was trying to implement the retry feature 😅

So, the `pin` method of the API simply sends the request to pin and returns a `true` or `false` if the request is accepted, and that's all I am checking to flag the pinning as a success: 

https://github.com/OriginProtocol/origin/blob/5fb19090e684c6f503b8ebd2e55c13377163d537/deployment/utils/gcf-pin-ipfs-hash/index.js#L33-L34

But, this doesn't mean the pinning was actually successful. There's a few other steps that the ipfs cluster service does to pin the data, the steps are outlined here: https://cluster.ipfs.io/documentation/internals/#pinning-an-item 

Referring to the link above, I am currently only returning after the 4th step highlighted in italics. So, right now I'm assuming all will be fine after the pinning request has been accepted.

The ideal way to check if the pinning was successful would be to call the `status` endpoint after pinning and checking if it returns a `pinned` status. The complexity with doing that is we don't know how long it will take to successfully pin. One solution would be something similar to how the retries would be implemented, to ping the `status` api a few times with increasing intervals and only flag it as pinned once a successful pinned status is returned. 

Let me know how that sounds and if you have any suggestions.

**Comment by franckc** [comment]
Hi @rohitverma007 - thanks for these additional details regarding pinning status. Very useful.

I'd be inclined to keep things simple for now and just test that the request was successful but not call the status api. Though maybe add a comment so that i the future we can come back to it and will have the context.

I could anticipate that as an extra measure in the future we add a separate "verification" cron script that runs periodically and rescans our entire corpus of listings/offers/users to verify their associated IPFS hashes are pinned. And that script would call the status endpoint.

Thanks again for being so detail oriented !

**Comment by rohitverma007** [comment]
@franckc 
Sounds good, and yea a separate cron script would be ideal in the future.

Should I still implement retrying the request? I imagine the main points of failure would actually be after  sending a successful request and before a successful pin. From the manual testing that I have done so far, one of the main reasons sending the request failed was due to an incorrect ipfshash and there isn't really a point retrying a request with an incorrect hash. Let me know your thoughts.

**Comment by franckc** [comment]
@rohitverma007 I think would be worth having a simple retry in the script to guard against the case where the IPFS server might be shortly down for maintenance or there is a network blip.

**Comment by rohitverma007** [comment]
> @rohitverma007 I think would be worth having a simple retry in the script to guard against the case where the IPFS cluster might be shortly down for maintenance or there is a network blip.

Done and pushed. I set the number of retries to 5, 500ms initial interval and interval doubles every retry. Let me know if that's alright.

**Pull request merged**